### ISSUE TITLE
Differences in SAC between LAC_TF2_GRAPH and SAC Haarnoja

### ISSUE THREAD

**Issue by rickstaa** [rickstaa/stable-learning-control#96]
This issue highlights some differences I found when comparing the [SAC implementation of Minghoa ](https://github.com/rickstaa/Actor-critic-with-stability-guarantee/blob/master/LAC/SAC_cost.py) with the implementation of [Haarnoja et al. 2019](https://github.com/rail-berkeley/softlearning).

## Different alpha optimization

During the comparison, I found that Minghoa optimizes the alpha Lagrange multiplier differently than Haarnoja et al. 2019.

### Haarnoja et al. 2019

Haarnoja et al. 2019 optimize alpha as follows:

1. They first calculate the alpha loss based on alpha (See [softlearning/algorithms/sac.py:L254-L255](https://github.com/rail-berkeley/softlearning/blob/491344805c3147678c46e92c9870de2ea910b197/softlearning/algorithms/sac.py#L254-L255)):

![image](https://user-images.githubusercontent.com/17570430/96748482-de11e100-13c9-11eb-992f-e14f7343a0f2.png)

2. Following they then optimize log_alpha based on this loss (See [softlearning/algorithms/sac.py:L262-L263](https://github.com/rail-berkeley/softlearning/blob/491344805c3147678c46e92c9870de2ea910b197/softlearning/algorithms/sac.py#L262-L263):

![image](https://user-images.githubusercontent.com/17570430/96748446-d3efe280-13c9-11eb-8f9c-5559ccbc512d.png)

3. Finally they then relate log_alpha to alpha.

### Minghoas version

Minghoa, on the other hand, does the following:

1. He calculates the alpha loss based on log_alpha:

https://github.com/rickstaa/LAC_TF2_TORCH_Translation/blob/8150ef056bcbf97d8651fbcaa1de0f91018b08c2/legacy/LAC_ORIGINAL/LAC_ORIGINAL/LAC/LAC_V1.py#L146-L148

Following he optimizes log_alpha based on this loss:

https://github.com/rickstaa/LAC_TF2_TORCH_Translation/blob/8150ef056bcbf97d8651fbcaa1de0f91018b08c2/legacy/LAC_ORIGINAL/LAC_ORIGINAL/LAC/LAC_V1.py#L149-L151

I verified this choice with Minghoa, and he said he copied it from Haarnoja. I, therefore, think he used an [old commit](https://github.com/rail-berkeley/softlearning/blob/3e781e668d36b364c320de62ff830f1ad7c7234c/softlearning/algorithms/sac.py#L209-L211) in which Haarnoja et al. 2019 were still using log_alpha in the loss function. They initially used log_alpha since it is more numerically stable (see [this issue](https://github.com/rail-berkeley/softlearning/issues/37)). They, however, reverted this change in [#a187972](https://github.com/rail-berkeley/softlearning/blob/a187972416694b53628eb5c2c1dcd9760f0c7b62/softlearning/algorithms/sac.py#L248) because it caused confusion. Reverting this, however, introduced a numerical instability (see [this issue](https://github.com/rail-berkeley/softlearning/issues/136)). This was eventually solved by using alpha in the loss function (to keep it in line with the article) but optimize log_alpha instead of alpha. The currently correct implementation should, therefore be:

![image](https://user-images.githubusercontent.com/17570430/96862073-495cc100-1465-11eb-9c85-0919343e0718.png)

@panweihit I went ahead and changed the code in the new versions I left the incorrect implementation in the legacy code as a reference. Related to this, I propose a change to the labda_loss. We use the following in our current implementation:

https://github.com/rickstaa/LAC_TF2_TORCH_Translation/blob/8150ef056bcbf97d8651fbcaa1de0f91018b08c2/legacy/LAC_ORIGINAL/LAC_ORIGINAL/LAC/LAC_V1.py#L145

It might, however, be more clear to use lambda instead of log_labda as this is more in line with the formulas depicted in [Han et al. 2020](https://arxiv.org/pdf/2004.14288.pdf).

![image](https://user-images.githubusercontent.com/17570430/96862896-604fe300-1466-11eb-8788-8f55fcac2af3.png)

While doing so, we, of course, have to make a quick note explaining why we optimize log_labda/log_alpha instead of lambda/alpha. We can then point to [the issue of Haarnoja et al. 2019](https://github.com/rail-berkeley/softlearning/issues/136).

## Different Critic optimization 

I found two differences in the way Haarnoja and Minghoa optimize the critic.

### Critic loss function

[Haarnoja et al 2019](https://github.com/rail-berkeley/softlearning/blob/cdf1e5890978ce2c487f3694dee26b9317592de2/softlearning/algorithms/sac.py#L197), [spinningup](https://github.com/openai/spinningup/blob/038665d62d569055401d91856abb287263096178/spinup/algos/tf1/sac/sac.py#L185) and [stable-baselines](https://github.com/hill-a/stable-baselines/blob/b3f414f4f2900403107357a2206f80868af16da3/stable_baselines/sac/sac.py#L223) use the following formula for calculating the critic loss:

![image](https://user-images.githubusercontent.com/17570430/96874055-fdfede80-1475-11eb-930e-b04c10adcac0.png)

Minghoa, however, omits the 0.5 multipliers when calculating the Critic loss (see [Actor-critic-with-stability-guarantee/LAC/SAC_cost.py#L147-L148](https://github.com/rickstaa/Actor-critic-with-stability-guarantee/blob/002c06289d8d16c6410b080322912eeae7f9acb8/LAC/SAC_cost.py#L147-L148
):

![image](https://user-images.githubusercontent.com/17570430/96926928-92893100-14b6-11eb-80e0-cb949db074d2.png)

I think omitting the 0.5 does not influence the result as it is mainly added to make the derivation easier (this is also explained in [this StackExchange question](https://datascience.stackexchange.com/questions/52157/why-do-we-have-to-divide-by-2-in-the-ml-squared-error-cost-function)). This can further be seen by the fact that the Spinningup authors include the 0.5 in their TensorFlow implementation but omit it in their PyTorch version. We are therefore free to include or exclude it. @panweihit maybe it is an idea to include it to prevents possible confusion as to why the code omits it while the multiplication factor is present in formula 7 of [Han et al. 2020](https://arxiv.org/pdf/2004.14288.pdf). Alternatively, I can also add a quick note on the topic while excluding it.

![image](https://user-images.githubusercontent.com/17570430/96874524-7c5b8080-1476-11eb-8aa5-c34c82a1544e.png)

### Different Critic bellman backup calculation

There is a difference in how Minghoa and Haarnoja et al. 2019 calculate the Critic target values.

![image](https://user-images.githubusercontent.com/17570430/97049898-b9aa3600-157c-11eb-9800-70c05f50673a.png)

Minghoa uses a target actor to calculate both the log_probabilities and target Q values (See [/LAC/SAC_cost.py#L138-L142](https://github.com/rickstaa/Actor-critic-with-stability-guarantee/blob/112a1721ce3dbdab755cd854ea2211c1dd969c44/LAC/SAC_cost.py#L138-L142), [/LAC/SAC_cost.py#L135](https://github.com/rickstaa/Actor-critic-with-stability-guarantee/blob/112a1721ce3dbdab755cd854ea2211c1dd969c44/LAC/SAC_cost.py#L135) and [/LAC/SAC_cost.py#L96](https://github.com/rickstaa/Actor-critic-with-stability-guarantee/blob/112a1721ce3dbdab755cd854ea2211c1dd969c44/LAC/SAC_cost.py#L96)). Haarnoja et al. 2019, stable-baselines and Spinning up don't have a target Actor and therefore use the main actor for calculating these values. This is in line with equation 6 of Haarnoja et al. 2019. I, therefore, think Minghoa re-used the entropy that is used in the LAC algorithm, in which we did need a target network when he was implementing the SAC algorithm. I do not know exactly if using the target Actor (which follows the main actor using an exponential moving average with a decay factor of 0.995)) does influence the results. On the one hand, it could cause the SAC algorithm to explore more and therefore get less stuck in local optimums. On the other hand, it could lead to lower data efficiency. In the end, however, both implementations should converge to the same result. @panweihit Let me know what you think. Nevertheless, I propose we stick with Haarnoja's implementation to keep our code consistent with the Literature.

## Different optimization order

The Optimization order differs between Minghoa and Haarnoja et al. 2019. 

### Minghoas optimization order:

In Minghoas version as he uses tf1.0 the optimization order is not defined (see [this question](https://stackoverflow.com/questions/43844510/is-session-runfetches-guaranteed-to-execute-its-fetches-arguments-in-order)) except for target update which is executed before optimizing the critic due to the specified control dependency (see [/LAC/SAC_cost.py#L136](https://github.com/rickstaa/Actor-critic-with-stability-guarantee/blob/112a1721ce3dbdab755cd854ea2211c1dd969c44/LAC/SAC_cost.py#L136). 

### Haarnoja et al. 2019

Harnoja et al. 2019 in their old code also don't define the optimization order. In the new code they, however, use the following order:

- Update Critic
- Update Actor
- Update alpha
- Update target network

In contrast to Minghoa Haarnoja therefore applies the target update after all the other parameters have been optimized (see [softlearning/algorithms/sac.py#L292-L298](https://github.com/rail-berkeley/softlearning/blob/cdf1e5890978ce2c487f3694dee26b9317592de2/softlearning/algorithms/sac.py#L292-L298).

![image](https://user-images.githubusercontent.com/17570430/96869023-3222d100-146f-11eb-9ff6-e0dd3b5e62ba.png)

If we look into other implementations, ([spinup/algos/tf1/sac/sac.py#L198-L205](https://github.com/openai/spinningup/blob/038665d62d569055401d91856abb287263096178/spinup/algos/tf1/sac/sac.py#L198-L205) and [stable_baselines/sac/sac.py#L465-L467](https://github.com/hill-a/stable-baselines/blob/b3f414f4f2900403107357a2206f80868af16da3/stable_baselines/sac/sac.py#L465-L467)) we see that they use the convention used by Haarnoja et al. 2019. I verified this with Minghoa, and he said the choice he made was arbitrary. @panweihit I, therefore, propose that we use the convention of Haarnoja et al. 2019 in the new code. I don't think changing the other would matter much, but I will do some quick tests.

### Actor loss function

Some SAC implementations use the min Q_target ([Minghoa's version](https://github.com/rickstaa/Actor-critic-with-stability-guarantee/blob/002c06289d8d16c6410b080322912eeae7f9acb8/LAC/SAC_cost.py#L126), [spinningup](https://github.com/openai/spinningup/blob/038665d62d569055401d91856abb287263096178/spinup/algos/pytorch/sac/sac.py#L216) and [Haarnoja et al.2019](https://github.com/rail-berkeley/softlearning/blob/cdf1e5890978ce2c487f3694dee26b9317592de2/softlearning/algorithms/sac.py#L227)) while calculating the actor loss while others ([stable-baselines](https://github.com/hill-a/stable-baselines/blob/b3f414f4f2900403107357a2206f80868af16da3/stable_baselines/sac/sac.py#L235-L236)) use for example Q1. This was discussed in[ this issue](https://github.com/hill-a/stable-baselines/issues/325) and did not influence the results. We can therefore stick to using the min Q_target as it is equal to the original implementation of Haarnoja et al. 2019.

### Different network output activation functions

In https://github.com/rickstaa/stable-learning-control/issues/97 I also found that there is a difference between Minghoa and the other implementations regarding the Gaussian actor and critic Output activation function.

## TODOS:

- [x] Validate whether changing log_alpha to alpha gives the same results.
    - [x] Doesn't seem to matter for the end result.
- [x] Validate whether changing the optimization order changes the result.
    - [x] Doesn't seem to matter for the end result.
- [x] Validate whether removing or adding the 0.5 constant does not influence the result.
    - [x] Doesn't seem to matter for the end result.
- [x] Validate whether using the target Actor for the entropy calculation changes the result.
    - [x] This does not influence the end result.
- [x] Validate whether using a different output activation function gives a different result.
    - [x] Doesn't seem to matter for the ability of the agent to train. Might still influence the performance.
